src/Common/Config/YAMLParser.cpp

```diff
@@ -0,0 +1,196 @@
+        if (key.starts_with(YAML_ATTRIBUTE_PREFIX))
+        {
+            /// we use substr(1) here to remove YAML_ATTRIBUTE_PREFIX from key
+            attr_vec.push_back(std::make_pair(key.substr(1), value));
+        }
+    }
+}
+
+template <typename T>
+bool YAMLParser::processNode(YAML::Node node, Poco::AutoPtr<Poco::XML::Document> xml, T * xml_node, attrs& attr_vec)
+{
+    attrs ret_vec;
+    for (auto it = node.begin(); it != node.end(); ++it)
+    {
+        auto key_node = it->first;
+        auto val_node = it->second;
+        if (key_node.IsNull())
+        {
+            /// yaml-cpp parser encounters a null node
+            continue;
+        }
+        else if (val_node.IsScalar())
+        {
+            /// YAML node like this
+            /// key: value
+            /// is parsed into
+            /// <key>value</key>
+            std::string key, value;
+            key = key_node.as<std::string>();
+            value = val_node.as<std::string>();
+            if (key.starts_with(YAML_ATTRIBUTE_PREFIX))
+            {
+                /// we use substr(1) here to remove YAML_ATTRIBUTE_PREFIX from key
+                attr_vec.push_back(std::make_pair(key.substr(1), value));
+            }
+            else
+            {
+                Poco::XML::Element * new_node = xml->createElement(key);
+                Poco::XML::Text * xml_value = xml->createTextNode(value);
+                new_node->appendChild(xml_value);
+                xml_node->appendChild(new_node);
+            }
+        }
+        else if (val_node.IsMap())
+        {
+            /// YAML node which contains other nodes
+            std::string key;
+            key = key_node.as<std::string>();
+            Poco::XML::Element * new_node = xml->createElement(key);
+            processNode(val_node, xml, new_node, ret_vec);
+            for (auto & iter : ret_vec)
+            {
+                new_node->setAttribute(iter.first, iter.second);
+            }
+            ret_vec.clear();
+            xml_node->appendChild(new_node);
+        }
+        else if (val_node.IsSequence())
+        {
+            /// YAML node which contains other nodes and/or values
+            std::string key, value;
+            key = key_node.as<std::string>();
+            /// Firstly we iterate through the sequence to get all xml attributes
+            for (const auto & seq_node : val_node)
+            {
+                /// if val_node is just a simple value node, it cannot contain attribute
+                /// attribute nodes consist of an attribute key and a value
+                /// so, in sequence with nodes ' - value ' and '- "@key":value' we want to check only the second one
+                if (!seq_node.IsScalar())
+                {
+                    seqNodeGetAttrs(seq_node, ret_vec);
+                }
+            }
+            /// Then we iterate through it again to convert all actual values to xml and set their attributes
```

Comment:
Comments make it better, but the function still look overcomplicated.

Reply:
I tried to rewrite this function to make it shorter:

```
void YAMLParser::processNode(const YAML::Node & node, Poco::XML::Element & parent_xml_element, Poco::XML::Document & xml_document)
{
    switch (node.Type())
    {
        case YAML::NodeType::Scalar:
        {
            auto value = node.as<String>();
            Poco::XML::Text * xml_value = xml_document.createTextNode(value);
            parent_xml_element.appendChild(xml_value);
            break;
        }

        case YAML::NodeType::Sequence:
        {
            for (const auto & child_node : node)
                processNode(child_node, parent_xml_element, xml_document);
            break;
        }

        case YAML::NodeType::Map:
        {
            for (const auto & key_value_pair : node)
            {
                const auto & key_node = key_value_pair.first;
                const auto & value_node = key_value_pair.second;
                auto key = key_node.as<String>();
                bool is_attribute = (key.starts_with(YAML_ATTRIBUTE_PREFIX) && value_node.IsScalar());
                if (is_attribute)
                {
                    auto attribute_name = key.substr(1);
                    auto value = value_node.as<String>();
                    parent_xml_element.setAttribute(attribute_name, value);
                }
                else
                {
                    Poco::XML::Element * xml_key = xml_document.createElement(key);
                    parent_xml_element.appendChild(xml_key);
                    processNode(value_node, *xml_key, xml_document);
                }
            }
            break;
        }

        case YAML::NodeType::Null: break;
        case YAML::NodeType::Undefined: break;
    }
}
```

I didn't check if it works, but here I had to write all that attribute-related logic only once. What do you think? Did I miss something?

Reply:
It seems that nodes like this:
```
sequence:
  - value1
  - value2
```
can be parsed incorrectly. I'll check if recursive call for processNode will parse values "without" key. Вut overall it really looks much simpler and clearer than before.

Reply:
If I undestood correctly the issue about sequences can be fixed like this:

```
case YAML::NodeType::Scalar:
{
    auto value = node.as<String>();
    Poco::XML::Text * xml_value = xml_document.createTextNode(value);
    if (hasChildTextNode(parent_xml_element))
    {
        /// We cannot add a second text node because ...
        cloneAsSibling(parent_xml_element)->appendChild(xml_value);
    }
    else
    {
        parent_xml_element.appendChild(xml_value);
    }
    break;
}
```

here `cloneAsSibling()` clones a xml node and append it to the parent of this xml node; and `hasChildTextNode()` checks if a xml node has a child of type `Poco::XML::Node::TEXT_NODE`.

Reply:
And it seems we don't need to pass `xml_document` to this function because it can be get anytime by calling `parent_xml_node.ownerDocument()`.